File: 0078-subsets/0078-subsets.java
class Solution {
        private static void f(int ind,int []arr,List<List<Integer>>ans,ArrayList<Integer>ds){
                if(ind>=arr.length){
                        ans.add(new ArrayList<>(ds));
                        return;
                }
                ds.add(arr[ind]);
                f(ind+1,arr,ans,ds);
                ds.remove(ds.size()-1);
                f(ind+1,arr,ans,ds);        
        }
    public List<List<Integer>> subsets(int[] arr) {
        List<List<Integer>>ans=new ArrayList<>();
            f(0,arr,ans,new ArrayList<>());
            return ans;
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void f(int i,vector<int>& arr,vector<int>& ds,vector<vector<int>>& ans){
        if(i>=arr.size()){
            ans.push_back(ds);
            return;
        }
        //include
        ds.push_back(arr[i]);
        f(i+1,arr,ds,ans);
        //exclude
        ds.pop_back();
        f(i+1,arr,ds,ans);
    }
    vector<vector<int>> subsets(vector<int>& arr) {
        vector<vector<int>>ans;
        vector<int>ds;
        f(0,arr,ds,ans);
        return ans;
    }
};